src/main/scala/org/constellation/snapshotstreaming/opensearch/mapper/SnapshotMapper.scala

```diff
@@ -118,24 +122,28 @@ abstract class SnapshotMapper[F[_]: Async, S <: OriginalSnapshot] {
   private def mapTransactionRef(nodeRef: OriginalTransactionReference): TransactionReference =
     TransactionReference(nodeRef.hash.value, nodeRef.ordinal.value)
 
-  def snapshotReferredBalancesInfo(
+  def balanceDiff(
     snapshot: S,
-    info: SnapshotInfo[_]
-  ): SortedMap[Address, Balance] = {
-    val snapshotReferredAddresses = extractSnapshotReferredAddresses(snapshot)
-    val bothAddresses = snapshotReferredAddresses.source ++ snapshotReferredAddresses.destination
-    val rewardsAddresses = fetchRewards(snapshot).toList.map(_.destination)
-    val addressesToKeep = bothAddresses ++ rewardsAddresses
-    val filteredBalances = info.balances.filter { case (address, _) => addressesToKeep.contains(address) }
-
-    val srcTransactions = snapshotReferredAddresses.source
-    val setZeroBalances =
-      (srcTransactions.toSet -- info.balances.keys.toSet)
-        .map(address => address -> Balance(0L))
-        .toSortedMap
-
-    filteredBalances ++ setZeroBalances
-  }
+    prevBalances: Option[SortedMap[Address, Balance]],
+    info: SnapshotInfo[_],
+  ): SortedMap[Address, Balance] =
+    prevBalances match {
+      case Some(prev) =>
+        val changed = info.balances.filterNot { case (address, balance) =>
+          prev.get(address).exists(_ === balance)
+        }
+
+        val explicitlyZeroed = {
+          val srcTransactions = extractSnapshotReferredAddresses(snapshot).source
+          (srcTransactions -- info.balances.keys.toSet)
```

Comment:
You don't need to construct a set
```suggestion
          (srcTransactions -- info.balances.keys)
```
Also, you can avoid unnecessary object creation by replacing `Balance(0L)` below with `Balance.empty`